Mathlib/RingTheory/Valuation/Basic.lean

```diff
@@ -472,29 +474,445 @@ theorem map {v' : Valuation R Γ₀} (f : Γ₀ →*₀ Γ'₀) (hf : Monotone f
+    let @MaybeDefEq.defEq u₃ _ _ _ d₂ ← isDefEqQ hΓ₂ mΓ₂ | return .none
+    let .some ⟨z, hxz⟩ ← mkAssociated v₁ v₂ h x | return .none
+    let .some ⟨w, hyw⟩ ← mkAssociated v₁ v₂ h y | return .none
+    trace[debug] m!"Transformed:\n({rel.toProp (α := Γ₁) q(inferInstance) x y})
+to:\n({rel.toProp (α := Γ₂) q(inferInstance) z w})"
+    match rel with
+    | .le => return .some ⟨q($z ≤ $w), q(Associated.le_iff_le (h := $h) $hxz $hyw)⟩
+    | .lt => return .some ⟨q($z < $w), q(Associated.lt_iff_lt (h := $h) $hxz $hyw)⟩
+    | .eq => return .some ⟨q($z = $w), q(Associated.eq_iff_eq (h := $h) $hxz $hyw)⟩
+    | .ne => return .some ⟨q($z ≠ $w), q(Associated.ne_iff_ne (h := $h) $hxz $hyw)⟩
+  | _, _ =>
+    let .some ⟨z, hxz⟩ ← mkAssociatedₘ v₁ v₂ h x | return .none
+    let .some ⟨w, hyw⟩ ← mkAssociatedₘ v₁ v₂ h y | return .none
+    trace[debug] m!"Transformed:\n({rel.toProp (α := Γ₁) q(inferInstance) x y})
+to:\n({rel.toProp (α := Γ₂) q(inferInstance) z w})"
+    match rel with
+    | .le => return .some ⟨q($z ≤ $w), q(Associatedₘ.le_iff_le (h := $h) $hxz $hyw)⟩
+    | .lt => return .some ⟨q($z < $w), q(Associatedₘ.lt_iff_lt (h := $h) $hxz $hyw)⟩
+    | .eq => return .some ⟨q($z = $w), q(Associatedₘ.eq_iff_eq (h := $h) $hxz $hyw)⟩
+    | .ne => return .some ⟨q($z ≠ $w), q(Associatedₘ.ne_iff_ne (h := $h) $hxz $hyw)⟩
+
+/-- Match the relation to be one of `≤`, `<`, `=`, or `≠`, and then use `mkProof` to build the
+proof wanted. -/
+def matchAndMkProof (h : Q(Valuation.IsEquiv $v₁ $v₂)) (e₁ : Q(Prop)) :
+    MetaM (Option ((e₂ : Q(Prop)) × Q($e₁ ↔ $e₂))) := do
+  match e₁ with
+  | ~q(@LE.le $Γ₁ $inst $a $b) => mkProof v₁ v₂ h .le a b
+  | ~q(@LT.lt $Γ₁ $inst $a $b) => mkProof v₁ v₂ h .lt a b
+  | ~q($a = $b) => mkProof v₁ v₂ h .eq a b
+  | ~q($a ≠ $b) => mkProof v₁ v₂ h .ne a b
+  | _ => return .none
+
+/-- The core simproc of `rw_equiv_tac`. Given `h : IsEquiv v₁ v₂`, find relations in `Γ₁` and
+transport them to `Γ₂`. -/
+def equivCore (h : Q(Valuation.IsEquiv $v₁ $v₂)) : Simp.Simproc := fun e : Expr ↦ do
+  match ← inferTypeQ e with
+  | ⟨1, ~q(Prop), e⟩ =>
+    match ← matchAndMkProof v₁ v₂ h e with
+    | .some ⟨e₂, pf⟩ => return .visit { expr := e₂, proof? := q(propext $pf) }
+    | .none => return .continue
+  | _ => return .continue
+
+/-- A tactic to rewrite expressions in a goal (e.g. `v₁ x ≤ 1`) with an equivalent one in the other
+value group or monoid (e.g. `v₂ x ≤ 1`), given `h : v₁.IsEquiv v₂`. -/
+elab "rw_val_equiv" equiv:(ppSpace colGt term:max) : tactic => do
+  let hE : Expr ← elabTerm equiv none
+  let ⟨0, h', h⟩ ← inferTypeQ hE | throwError "given term is not a proof"
+  match h' with
+  | ~q(@Valuation.IsEquiv $R $Γ₁ $Γ₂ $hR $hΓ₁ $hΓ₂ $v₁ $v₂) =>
+      liftMetaTactic1 fun e ↦ do
+        let target ← instantiateMVars (← e.getType)
+        let ctx ← Simp.mkContext (simpTheorems := #[])
+        let (r, _) ← Simp.mainCore target ctx (methods := {post := equivCore v₁ v₂ h})
+        let i ← applySimpResultToTarget e target r
```

Comment:
afaik, this is a new pattern, perhaps a few comments explaining what's going on could make sense